executor/distsql.go

```diff
@@ -386,6 +389,38 @@ func (e *IndexLookUpExecutor) Open(ctx context.Context) error {
 	return err
 }
 
+func (e *IndexLookUpExecutor) buildTableKeyRanges() (err error) {
+	sc := e.ctx.GetSessionVars().StmtCtx
+	if e.partitionTableMode {
+		if e.keepOrder { // this case should be prevented by the optimizer
+			return errors.New("invalid execution plan: cannot keep order when accessing a partition table by IndexLookUpReader")
+		}
+		e.feedback.Invalidate() // feedback for partition tables is not ready
+		e.partitionKVRanges = make([][]kv.KeyRange, 0, len(e.prunedPartitions))
+		for _, p := range e.prunedPartitions {
```

Comment:
Some unnecessary key ranges are generated here.
The key ranges is constructed by 2 dimensions:  partitions * ranges, but not every partition need to combine every range. For example:
p1  + r1 r2
p2  + r3
they are the necessary ones.

But the generated `partitionKVRanges` is:
p1 + r1 r2 r3
p2 + r1 r2 r3

That problem is similar to the partition pruning problem: without that optimization, the code still work.

Reply:
I suggest adding some comment here so if someone who read the code later on would be aware of the possible optimization.

Reply:
It seems you are right. 
To solve this problem, we can prune and adjust `e.ranges` here according to `e.prunedPartitions`.
But it seems a little complex, so I will leave some comments here and let it as a further optimization.

Reply:
It is updated. Some comments are added.